storage: fix TestTxnObeysLeaseExpiration

vivekmenezes opened the pull request:
Return an error with the transaction in the PENDING state
when a transaction doesn't meet its deadline. This allows
the client to issue a Rollback() that cleans up the transaction
intents. This mechanism is a bit unreliable because if the client
fails before issuing a Rollback(), the transaction remains pending
with the intents alive. But it is deemed to be a good solution for
the deadline exceeds case, which is a rare event. A rare failure
followed by a rare client failure will result in intents lying
around until the transaction record itself expires.

fixes #7031

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9906)

<!-- Reviewable:end -->

vivekmenezes commented:
Looks like some tests are failing. Sorry, hold off from reviewing.

vivekmenezes commented:
Any thoughts on this?

tschottdorf commented:
My apologies @vivekmenezes, this totally fell off my radar.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease_test.go, line 604 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURN_nBHBnezfz-7A__:-KURN_nCDV3heGKZzmOF:bf2v54) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/sql/lease_test.go#L604)):*
> ```Go
> 	defer s.Stopper().Stop()
> 	// TODO(vivek): remove this sleep in this PR. Sleep because the server
> 	// clock is set forward when we configure the MaxOffset.
> ```

What's happening here? Whatever is going on shouldn't, I think.

---

*[pkg/storage/replica_command.go, line 590 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURNhypH1_XK6vP8q_L:-KURNhypH1_XK6vP8q_M:b-8eepi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/storage/replica_command.go#L590)):*
> ```Go
> 		// issue a Rollback() for cleanup.
> 		return reply,
> 			// TODO(vivek): returning nil here is not breaking any of the tests.
> ```

Why not add a test case while you're here?

---

*[pkg/storage/replica_command.go, line 592 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURNnIoDuGUeTBA8dWK:-KURNnIoDuGUeTBA8dWL:b-7bhwyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/storage/replica_command.go#L592)):*
> ```Go
> 			// TODO(vivek): returning nil here is not breaking any of the tests.
> 			intentsToProposalData(roachpb.AsIntents(args.IntentSpans, reply.Txn), &args),
> 			roachpb.NewTransactionStatusError("transaction deadline exceeded")
> ```

As far as I can tell, SQL is the only user of this. This doesn't impact its use case?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KURO2sKg6cG8-R-ITMv:b187lf7)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease_test.go, line 604 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURN_nBHBnezfz-7A__:-KU_IDBKKJ7CJsFzKAvM:bysfjf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/sql/lease_test.go#L604)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's happening here? Whatever is going on shouldn't, I think.
</details>
This will be removed after #9504 is merged

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease_test.go, line 604 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURN_nBHBnezfz-7A__:-KU_IOxI7XHJzcLHnJwR:b-2cb0si) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/sql/lease_test.go#L604)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> This will be removed after #9504 is merged
</details>
I'll wait for that PR to merge before rebasing. Hold on until then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Ready for another review. Thanks!

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/lease_test.go, line 604 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURN_nBHBnezfz-7A__:-KUtL_4Nlljkd9H2pdiV:bp11ciy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/sql/lease_test.go#L604)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I'll wait for that PR to merge before rebasing. Hold on until then.
</details>
#9504 is merged, I've removed the Sleep()

---

*[pkg/storage/replica_command.go, line 590 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURNhypH1_XK6vP8q_L:-KUtLs3bjxUPikmJejqK:b7swnri) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/storage/replica_command.go#L590)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why not add a test case while you're here?
</details>
This code returns the generic ProposalData{} which is returned with every error. I don't see the need to test anything specific here.

---

*[pkg/storage/replica_command.go, line 592 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURNnIoDuGUeTBA8dWK:-KUtM7tiLvrWutQTtXYk:bhxipz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/storage/replica_command.go#L592)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> As far as I can tell, SQL is the only user of this. This doesn't impact its use case?
</details>
SQL is the only user. As long as SQL keeps extending the transaction deadline it will not fail. This is only to guarantee that an expired table descriptor lease is not used by the KV  layer to commit a transaction that is created out of an old table descriptor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KUtN7-QQXX3uJmw1_kv:buswtg8)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Oops this is still failing stress tests :-1: 

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KUtO5PQ0TYxK3xJEpmQ:b-u5rnwn)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
This is ready for another review. I've had to remove the call to cleanupTxnLocked() whenever a txn errors out with TransactionStatusError. This is primarily because cleanupTxnLocked() doesn't cleanup any dangling intents for the txn. All TransactionStatusError cases rely on a higher layer call to txn.Rollback() which will cleanup the intents. I'd like to hear what you think? Thanks

tamird commented:
@andreimatei should take a look. You also seem to have checked-in `build/npm.installed`, which should be removed.

---

Reviewed 5 of 6 files at r2.
Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KUwv2kUmh3xIpEmm_F0:bdncj0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
> ```Go
> 	case *roachpb.TransactionStatusError:
> 		// Likely already committed or more obscure errors such as epoch or
> 		// timestamp regressions; consider txn dead. Do not cleaup txn because
> ```

"consider txn dead" is now unclear. perhaps remove it. Also, there are two distinct comments here, and the latter would typically begin with "NB:" in such cases.

---

*[pkg/kv/txn_coord_sender_test.go, line 537 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwvGzO6Kdm-MaZkGuI:-KUwvGzPUpm6oZ-SWdCp:bj8gb5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender_test.go#L537)):*
> ```Go
> }
> 
> func assertTransactionDeadlineError(t *testing.T, e error) {
> ```

please don't do this. This helper will cause test failures to report an unhelpful wrong line number.

In general, test helpers that call `t.{Error,Fatal}` are a bad idea for this reason.

---

*[pkg/sql/lease_test.go, line 601 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwvfCOQLkT8tIRDvLM:-KUwvfCOQLkT8tIRDvLN:b-q388fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/sql/lease_test.go#L601)):*
> ```Go
> 
> 	params, _ := createTestServerParams()
> 	// Increase the MaxOffset so that the clock can be updated to expire the
> ```

Why is it 10*csql.LeaseDuration? Such a specific value invites speculation; how about you set it to `math.MaxInt64`?

---

*[pkg/sql/lease_test.go, line 624 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwvpWtqIhxjaKF895U:-KUwvpWtqIhxjaKF895V:b-1nh511) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/sql/lease_test.go#L624)):*
> ```Go
> }
> 
> func runCommandAndExpireLease(t *testing.T, clock *hlc.Clock, sqlDB *gosql.DB, sql string) {
> ```

as I mentioned above, this test helper produces suboptimal error messages, but I won't insist on this.

---

*[pkg/sql/lease_test.go, line 625 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUww0-VPeWI2xA-s53P:-KUww0-VPeWI2xA-s53Q:bik8gr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/sql/lease_test.go#L625)):*
> ```Go
> 
> func runCommandAndExpireLease(t *testing.T, clock *hlc.Clock, sqlDB *gosql.DB, sql string) {
> 	t.Logf("now: %s, execute sql: %s", clock.Now(), sql)
> ```

did you mean to leave this in?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KUwwTfNLGK8rCX7mKgY:b-m1ykhv)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KUx8KolYz1d_tVIxrEi:bpntw7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "consider txn dead" is now unclear. perhaps remove it. Also, there are two distinct comments here, and the latter would typically begin with "NB:" in such cases.
</details>
Reworded this comment

---

*[pkg/kv/txn_coord_sender_test.go, line 537 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwvGzO6Kdm-MaZkGuI:-KUx8qs3wHbAlb5V7TUn:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender_test.go#L537)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> please don't do this. This helper will cause test failures to report an unhelpful wrong line number.
> 
> In general, test helpers that call `t.{Error,Fatal}` are a bad idea for this reason.
</details>
removed

---

*[pkg/sql/lease_test.go, line 604 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KURN_nBHBnezfz-7A__:-KUxA7OTTKiAtpnRZc7J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8b2ac60258120b9e538cb581ff9b21f8aad213e/pkg/sql/lease_test.go#L604)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> #9504 is merged, I've removed the Sleep()
</details>
Done.

---

*[pkg/sql/lease_test.go, line 601 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwvfCOQLkT8tIRDvLM:-KUxBKzAcXQmlNeQE_U-:b-izgvtk) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/sql/lease_test.go#L601)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Why is it 10*csql.LeaseDuration? Such a specific value invites speculation; how about you set it to `math.MaxInt64`?
</details>
changed to use math.MaxInt64

---

*[pkg/sql/lease_test.go, line 624 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwvpWtqIhxjaKF895U:-KUx9z5fUfsXpa3N1uQM:b-irrgzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/sql/lease_test.go#L624)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> as I mentioned above, this test helper produces suboptimal error messages, but I won't insist on this.
</details>
This helper will log line numbers from the helper which is okay for this test.

---

*[pkg/sql/lease_test.go, line 625 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUww0-VPeWI2xA-s53P:-KUx9MrN49HXZ3MxgybX:b-jpkjkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/sql/lease_test.go#L625)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to leave this in?
</details>
Removed, and added the "sql" text to the other t.Fatal() errors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KUxD7exJXA8-Auz18ac:bzpibw) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Reworded this comment
</details>
I still think this comment is lacking. Why is `cleanupTxnLocked` incapable of cleaning up intents? Where is the higher layer going to issue a rollback? This comment invites me to go spelunking around the code and also makes it difficult to verify that things still work as the author (i.e. you) expected at the time of writing.

---

*[pkg/sql/lease_test.go, line 637 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUxCnLsiBSEhhBwtJ2v:-KUxCnLt3dyjflsfC893:b43an5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f327f7a282f58c4386f969e98320f29b7baedd27/pkg/sql/lease_test.go#L637)):*
> ```Go
> 	}
> 	if _, err := txn.Exec(sql); err != nil {
> 		t.Fatalf("%s, err = %v", sql, err)
> ```

s/err = %v/err = %s/ since `err` can't be nil.

---

*[pkg/sql/lease_test.go, line 645 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUxD-PvpSSSSoDyojA6:-KUxD-PvpSSSSoDyojA7:b43an5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f327f7a282f58c4386f969e98320f29b7baedd27/pkg/sql/lease_test.go#L645)):*
> ```Go
> 	// Run another transaction that pushes the above transaction.
> 	if _, err := sqlDB.Query("SELECT * FROM t.kv"); err != nil {
> 		t.Fatalf("%s, err = %v", sql, err)
> ```

s/err = %v/err = %s/ since `err` can't be nil.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 6 files at r2, 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease_test.go, line 637 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUxCnLsiBSEhhBwtJ2v:-KUywc2_cBe1DgoWlCfE:b2fqnvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f327f7a282f58c4386f969e98320f29b7baedd27/pkg/sql/lease_test.go#L637)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/err = %v/err = %s/ since `err` can't be nil.
</details>
Why care about this? I think both `%v` and `%s` should be acceptable for formatting errors. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KUyxB20X5npoJpfLZ-u:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
To be honest, it seems to me like we oughta be able to do better then say "hey client, we got into an unrecoverable state over here, mind lending a hand with the cleanup?". But I don't really know what I'm talking about so copying Ben's LGTM.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KUz0GoV6Ibfh-T1MSm2:bzhvnqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I still think this comment is lacking. Why is `cleanupTxnLocked` incapable of cleaning up intents? Where is the higher layer going to issue a rollback? This comment invites me to go spelunking around the code and also makes it difficult to verify that things still work as the author (i.e. you) expected at the time of writing.
</details>
+1. The reason why we're incapable of cleaning up the intents (which I guess is explained in #3037) is pretty obscure. In fact, so obscure that it makes me think that we could probably do something about it? The bug says we already do something similar for `TransactionRetryError`; maybe we can do the same for this one, or invent a more specific error for which we do it?

Another related question I have is how come we're relying on the client and not sending the rollback `EndTransaction` ourselves right here? I think Ben mentioned that too in #7031. It seems the natural thing to do to me, short of actually making `cleanupTxnLocked()`. In fact I guess you can send the `EndTransaction` async and still return the error immediately to the client. Is there difficulty I'm not seeing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KUz4D060hFuEndE_Cbl:b-p2bvom)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
This PR has enough cooks, unassigning.

vivekmenezes commented:
Tobi,

Sorry about that. I've needed to rework it a bit and I'm still not happy
where it is. I'll send out an update hopefully tomorrow and you can comment
if you'd like. Thanks

On Mon, Oct 31, 2016, 8:38 PM Tobias Schottdorf <notifications@github.com>
wrote:

> This PR has enough cooks, unassigning.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9906#issuecomment-257461920>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBOHuTLzJ6R6McQ7Vr69dwd7_1Hjcks5q5ooLgaJpZM4KUxjX>
> .
>

vivekmenezes committed:
storage: fix TestTxnObeysLeaseExpiration

Return a TransactionStatusError with the transaction in the
PENDING state when a transaction doesn't meet its deadline.
The deadline is checked while evaluating an EndTransactionRequest.
The client on seeing an error calls Txn.CleanupOnError()
which cleans up the transaction and its intents.

Removed the need to cleanup a transaction on the txn coordinator
when it fails with the TransactionStatusError to allow it to be
references later when Txn.CleanupOnError() is called.

fixes #7031

vivekmenezes commented:
I've had to significantly change the test to fit with the world where MaxOffset is not encouraged to be modified in tests. It's a lot simpler. This PR still relies on the SQL layer calling CleanupOnError() on the transaction. Thanks for reviewing.

---

Review status: 1 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVaiROHJC78wtecx1Vj:bcsnm5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> +1. The reason why we're incapable of cleaning up the intents (which I guess is explained in #3037) is pretty obscure. In fact, so obscure that it makes me think that we could probably do something about it? The bug says we already do something similar for `TransactionRetryError`; maybe we can do the same for this one, or invent a more specific error for which we do it?
> 
> Another related question I have is how come we're relying on the client and not sending the rollback `EndTransaction` ourselves right here? I think Ben mentioned that too in #7031. It seems the natural thing to do to me, short of actually making `cleanupTxnLocked()`. In fact I guess you can send the `EndTransaction` async and still return the error immediately to the client. Is there difficulty I'm not seeing?
</details>
I've tried all sorts of ways to cleanup the intents from either the replica itself or from the txn coordinator and was unable to do so in a satisfactory way.  One of the motivations for doing it this way is that the sql layer does call CleanupOnError() for non retryable transaction errors so we already have a cleanup mechanism implemented at a higher layer; why do it again? I've burnt about 2 days going back and forth on one implementation after another and I'm not confident any of them are pretty enough or work correctly. The best alternative I came up with was using a synchronous version of tryAsyncAbort(), but I was not convinced that it was better. The client will still call a redundant CleanupOnError().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KVb0E4uZPiIZlIw5fIA:b-qw52k3)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -78,9 +80,12 @@ func (s *LeaseState) Expiration() time.Time {
+// hasSomeLifeLeft returns true if the lease has at least a minimum of
+// lifetime left until expiration, and thus can be used.
 func (s *LeaseState) hasSomeLifeLeft(clock *hlc.Clock) bool {
+	if s.testingKnobs.CanUseExpiredLeases {
```
FYI, I can also set MinLeaseDuration to a large negative number to achieve the same.

tamird commented:
FYI I've removed `SetMaxOffset` entirely, so this may need a rebase.

Looks good to me but definitely @andreimatei should sign off.

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVb2shAyqesu3F7W_FQ:bqr1s7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
The discussion y'all are having aside, this comment is still lacking for the reasons I originally mentioned:

>  Where is the higher layer going to issue a rollback? This comment invites me to go spelunking around the code and also makes it difficult to verify that things still work as the author (i.e. you) expected at the time of writing.

---

*[pkg/sql/lease_test.go, line 611 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVb39xlnDXL-gNFMHbT:-KVb39xlnDXL-gNFMHbU:b-tq3amf) ([raw file](https://github.com/cockroachdb/cockroach/blob/431e54725309d3f60ea1dc6ef5df9acd3d2e19be/pkg/sql/lease_test.go#L611)):*
> ```Go
> 	}
> 
> 	// The above INSERT has created a lease on the table; expire the lease.
> ```

s/created/acquired/

---

*[pkg/sql/lease_test.go, line 615 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVb3G_2pTTfMmpUq2mt:-KVb3G_2pTTfMmpUq2mu:b-wiwfrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/431e54725309d3f60ea1dc6ef5df9acd3d2e19be/pkg/sql/lease_test.go#L615)):*
> ```Go
> 	// because CanUseExpiredLeases is set. The SQL commands will set the
> 	// transaction deadline to the lease expiration time (in the past), and
> 	// the transaction will eventually fail with a deadline exceeded error.
> ```

woooh, that's quite a hack!

---

*[pkg/storage/replica_command.go, line 600 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVb3KgekJu5O-lB7gnk:-KVb3KgekJu5O-lB7gnl:byi4gez) ([raw file](https://github.com/cockroachdb/cockroach/blob/431e54725309d3f60ea1dc6ef5df9acd3d2e19be/pkg/storage/replica_command.go#L600)):*
> ```Go
> 		return reply, ProposalData{}, roachpb.NewTransactionStatusError(
> 			"transaction deadline exceeded")
> ```

errant newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KVb2RRep-T3SJZxgFfo:blcn215)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 3 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVb6pFYAoAa4e-kVQkr:b-3hqe6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> The discussion y'all are having aside, this comment is still lacking for the reasons I originally mentioned:
> 
> >  Where is the higher layer going to issue a rollback? This comment invites me to go spelunking around the code and also makes it difficult to verify that things still work as the author (i.e. you) expected at the time of writing.
</details>
Is the comment any better?  I've also added a comment to sql/session.go

---

*[pkg/sql/lease_test.go, line 611 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVb39xlnDXL-gNFMHbT:-KVb5DG2DbVUVZY86dXI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/431e54725309d3f60ea1dc6ef5df9acd3d2e19be/pkg/sql/lease_test.go#L611)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/created/acquired/
</details>
Done.

---

*[pkg/sql/lease_test.go, line 615 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVb3G_2pTTfMmpUq2mt:-KVb5FSfu6lC59ef3nsq:bpg221n) ([raw file](https://github.com/cockroachdb/cockroach/blob/431e54725309d3f60ea1dc6ef5df9acd3d2e19be/pkg/sql/lease_test.go#L615)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> woooh, that's quite a hack!
</details>
It's triggered by a knob. I originally was setting MinLeaseDuration to a large negative number to actuate the same behavior, but I thought the use of a knob CanUseExpiredLeases makes the hack more explicit 

---

*[pkg/storage/replica_command.go, line 600 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVb3KgekJu5O-lB7gnk:-KVb4pdg0itHzFYzgzNr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/431e54725309d3f60ea1dc6ef5df9acd3d2e19be/pkg/storage/replica_command.go#L600)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> errant newline
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 7 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVb4Sn9LsDM__3Rp8ah:bvfexb) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Is the comment any better?  I've also added a comment to sql/session.go
</details>
(how about we drop this `case *roachpb.TransactionStatusError` altogether? What's the point for it? This `switch` is not handling all possible errors, and the comment you're adding seems to me that it applies equally well to everything else.)

I have no particularly strong feelings as I don't really know the code involved. I'm sure the situation is complex :) So feel free to ignore.
The reason for "doing the cleanup it again" is the logical separation of the client and the "server". In some situations, the server cannot automatically cleanup after an error, as it doesn't know if the client still wants the transaction or not. In other cases, the server knows that a transaction cannot possibly be used at all any more, and then it tries to cleanup (at least I *think* that the server generally tries to cleanup early). The point is, I guess, that you don't want to rely on the client too much. And, technically, saving some time. For example, you can rely on the sql executor to not drop errors on the floor and to do cleanup, but what about other direct users of `client.Txn`? Will they surely do the cleanup? If they just call `txn.Commit()` and ignore the error, there's gonna be no cleanup (right?).

BTW, in recent conversations around watercoolers it has become apparent (to me at least, maybe it was apparent to other people for a long time) that the `TxnCoordSender` should be considered part of the "client" (in fact it should be fused with `client.Txn`). So doing cleanup here, is not, in my eyes, much better than not doing it at all (or at least not conceptually better, although it'd still have the practical advantage that the user of `client.Txn` cannot ignore the error). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 5 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVdYKcJqc63ZGah86jj:bmvcfls) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> (how about we drop this `case *roachpb.TransactionStatusError` altogether? What's the point for it? This `switch` is not handling all possible errors, and the comment you're adding seems to me that it applies equally well to everything else.)
> 
> I have no particularly strong feelings as I don't really know the code involved. I'm sure the situation is complex :) So feel free to ignore.
> The reason for "doing the cleanup it again" is the logical separation of the client and the "server". In some situations, the server cannot automatically cleanup after an error, as it doesn't know if the client still wants the transaction or not. In other cases, the server knows that a transaction cannot possibly be used at all any more, and then it tries to cleanup (at least I *think* that the server generally tries to cleanup early). The point is, I guess, that you don't want to rely on the client too much. And, technically, saving some time. For example, you can rely on the sql executor to not drop errors on the floor and to do cleanup, but what about other direct users of `client.Txn`? Will they surely do the cleanup? If they just call `txn.Commit()` and ignore the error, there's gonna be no cleanup (right?).
> 
> BTW, in recent conversations around watercoolers it has become apparent (to me at least, maybe it was apparent to other people for a long time) that the `TxnCoordSender` should be considered part of the "client" (in fact it should be fused with `client.Txn`). So doing cleanup here, is not, in my eyes, much better than not doing it at all (or at least not conceptually better, although it'd still have the practical advantage that the user of `client.Txn` cannot ignore the error). 
> 
</details>
Yeah, if we're not doing anything in this `case` we can just remove it and this whole comment. 

I think I'd like to preserve the client.Txn/kv.TxnCoordSender split. Even though our client.Txns are currently always local to their TxnCoordSender, I don't know if that's always going to be the case in the future. To me TxnCoordSender belongs closer to DistSender than to client.Txn.

---

*[pkg/sql/session.go, line 402 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVdZOGc7Zz0WCnLw2NU:-KVdZOGdMDnensmal7wD:b-gx5fsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/002071ede6b1b85a8341c51c7feb7ae3c2238d06/pkg/sql/session.go#L402)):*
> ```Go
> 		e.TxnAbortCount.Inc(1)
> 		// This call cleans up the transaction and all its intents for errors
> 		// like TransactionStatusError.
> ```

Why call out TransactionStatusError specifically here? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-:-KVdabEWVs3RYZOPA0Lx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVduS8g9Z7RUMaxw9y1:b-g2z42y) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, if we're not doing anything in this `case` we can just remove it and this whole comment. 
> 
> I think I'd like to preserve the client.Txn/kv.TxnCoordSender split. Even though our client.Txns are currently always local to their TxnCoordSender, I don't know if that's always going to be the case in the future. To me TxnCoordSender belongs closer to DistSender than to client.Txn.
</details>
Do you have any arguments for that @bdarnell? Having clients not colocated with their TxnCoordSender in the future is the main reason for making the clients fatter (i.e. merge TxnCoordSender into `internal/client`). If you think of future DistSQL clients which carry out parallel operations in the cluster, splitting the client state artificially between one coordinating node and the actual client severely complicates things, which we already experience today with `TxnCoordSender`'s heartbeat goroutine (or more generally, state).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

tamird commented:
I'm going to file an issue to fatten the client in this way; let's leave
this discussion to that issue instead of this PR.

On Nov 3, 2016 07:21, "Tobias Schottdorf" <notifications@github.com> wrote:

> Review status: all files reviewed at latest revision, 5 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/kv/txn_coord_sender.go, line 805 at r2
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVduS8g9Z7RUMaxw9y1:b-g2z42y>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805>):*
> *Previously, bdarnell (Ben Darnell) wrote…*
>
> Yeah, if we're not doing anything in this case we can just remove it and
> this whole comment.
>
> I think I'd like to preserve the client.Txn/kv.TxnCoordSender split. Even
> though our client.Txns are currently always local to their TxnCoordSender,
> I don't know if that's always going to be the case in the future. To me
> TxnCoordSender belongs closer to DistSender than to client.Txn.
>
>
> Do you have any arguments for that @bdarnell <https://github.com/bdarnell>?
> Having clients not colocated with their TxnCoordSender in the future is the
> main reason for making the clients fatter (i.e. merge TxnCoordSender into
> internal/client). If you think of future DistSQL clients which carry out
> parallel operations in the cluster, splitting the client state artificially
> between one coordinating node and the actual client severely complicates
> things, which we already experience today with TxnCoordSender's heartbeat
> goroutine (or more generally, state).
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9906>*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9906#issuecomment-258117407>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPPjIlnjwds2jXk7twzHoVDdga_NDks5q6cPLgaJpZM4KUxjX>
> .
>

vivekmenezes commented:
Review status: 3 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVeXd8djAaRPzWH_SQd:b-9ogegh) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Do you have any arguments for that @bdarnell? Having clients not colocated with their TxnCoordSender in the future is the main reason for making the clients fatter (i.e. merge TxnCoordSender into `internal/client`). If you think of future DistSQL clients which carry out parallel operations in the cluster, splitting the client state artificially between one coordinating node and the actual client severely complicates things, which we already experience today with `TxnCoordSender`'s heartbeat goroutine (or more generally, state).
</details>
I've removed this case and moved the comment down to the default case and removed the older comment for the default case along with the TODO questioning its truth, that was wrong.

---

*[pkg/sql/session.go, line 402 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVdZOGc7Zz0WCnLw2NU:-KVeXUIqNk2LMazI0wwE:b972dyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/002071ede6b1b85a8341c51c7feb7ae3c2238d06/pkg/sql/session.go#L402)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why call out TransactionStatusError specifically here? 
</details>
changed this comment to say that transactions that are PENDING are rolled back via this call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 848 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVecgTvY8IxNhw2SACD:-KVecgTw0-da8UdS6ZSl:b-9yvup) ([raw file](https://github.com/cockroachdb/cockroach/blob/0458f940ba1031a93ff16a3c9987008a47076cc4/pkg/kv/txn_coord_sender.go#L848)):*
> ```Go
> 		// can come from a unique index violation.
> 		//
> 		// TODO(bdarnell): Is this valid? Unless there is a single CPut in
> ```

how come you removed this TODO?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 6 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/kv/txn_coord_sender.go, line 848 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVecgTvY8IxNhw2SACD:-KVedg1cGr5OKu7rJD8M:b-26fate) ([raw file](https://github.com/cockroachdb/cockroach/blob/0458f940ba1031a93ff16a3c9987008a47076cc4/pkg/kv/txn_coord_sender.go#L848)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come you removed this TODO?
</details>
I didn't see a point to it. There are three places where we handle ConditionFailedError in the SQL layer, and in all three cases we rollback the transaction. I've gone ahead and added ConditionFailedError to the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 838 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVehEELwmvqIaumlfwr:-KVehEELwmvqIaumlfws:b-8kj5ml) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bcd9ff1fb31a2725a84fd54eeef14ecf3cf8d7a/pkg/kv/txn_coord_sender.go#L838)):*
> ```Go
> 	default:
> 		// Do not clean up the transaction since the client might still want
> 		// to continue the transaction. For example, on seeing an error such
> ```

Remove "might still want to continue". I'm pretty sure the transaction can't be continued, but we are leaving cancellation up to the client anyway. 

---

*[pkg/kv/txn_coord_sender.go, line 848 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVecgTvY8IxNhw2SACD:-KVehQd2Bu6q4eApiwRQ:b-49e4zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0458f940ba1031a93ff16a3c9987008a47076cc4/pkg/kv/txn_coord_sender.go#L848)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I didn't see a point to it. There are three places where we handle ConditionFailedError in the SQL layer, and in all three cases we rollback the transaction. I've gone ahead and added ConditionFailedError to the comment.
</details>
This TODO was based on the assumption that continuing after a ConditionFailedError was the main reason we wouldn't want to clean up the transaction in the TxnCoordSender. Now that we have other reasons for not cleaning up here, the question of whether the transaction can be continued is moot. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVelqxE_I366elYn2E2:b-gjk4o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I've removed this case and moved the comment down to the default case and removed the older comment for the default case along with the TODO questioning its truth, that was wrong.
</details>
I'd like to verify my understanding here, please:
Maybe part of my rant above on client/server split and how the server should be as good as possible about cleaning up is unfounded - `TxnCoordSender` is the only guy that has a global view of a transaction's intents. So the only cleanup that could possibly be done below the `TxnCoordSender` is cleanup of intents **local to the range** that the transaction happened to be touching when the "deadline exceeded" condition has been encountered. Is that right?
I guess in practice that's still relevant, since transactions I guess are frequently local to a single range, but generally the `TxnCoordSender` and/or the `client.Txn` above it need to be involved in the cleanup.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 838 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KVehEELwmvqIaumlfwr:-KVeqIorA9RsAYI5P7OC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bcd9ff1fb31a2725a84fd54eeef14ecf3cf8d7a/pkg/kv/txn_coord_sender.go#L838)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Remove "might still want to continue". I'm pretty sure the transaction can't be continued, but we are leaving cancellation up to the client anyway. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906#-KUwv2kUmh3xIpEmm_F-:-KVh5dM-REsJuyTpqSTh:b-v75vqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/715881b899e2935e93173d2f9e9bb2dca67a5c73/pkg/kv/txn_coord_sender.go#L805)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I'd like to verify my understanding here, please:
> Maybe part of my rant above on client/server split and how the server should be as good as possible about cleaning up is unfounded - `TxnCoordSender` is the only guy that has a global view of a transaction's intents. So the only cleanup that could possibly be done below the `TxnCoordSender` is cleanup of intents **local to the range** that the transaction happened to be touching when the "deadline exceeded" condition has been encountered. Is that right?
> I guess in practice that's still relevant, since transactions I guess are frequently local to a single range, but generally the `TxnCoordSender` and/or the `client.Txn` above it need to be involved in the cleanup.
</details>
There are two things to clean up: the intents and the transaction record itself. The transaction record is the more important part (intents will be cleaned up lazily but relatively quickly if the transaction record is updated, while a transaction record that is left PENDING has to time out before any progress can be made). The transaction record is not necessarily on the node where the error was discovered, so any cleanup has to be upstream of the DistSender (unless the error happens to be on the right range to begin with, but then we run into the problem that we can't write and return an error at the same time)

My thinking about the client.Txn/TxnCoordSender split is not about distsql, but about the vague possibility that we might reintroduce a direct KV api or other non-sql interface, and we'd want a thin client layer separate from TxnCoordSender. But that's pretty speculative at this point so I'd be open to combining them if it simplifies things. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9906)*
<!-- Sent from Reviewable.io -->
